### ISSUE TITLE
memory leak in dense x csc multiplication

### ISSUE THREAD

**Issue by simonwm** [flatironinstitute/sparse_dot#16]
Thanks again for this amazing package!

This time I found a memory leak for a specific matrix format combination:

```import scipy.sparse
import numpy as np
import sparse_dot_mkl
import memory_profiler
import time

nA = 10
nB = 1000
formats = ['dense','csr','csc']
for first_format in formats:
    for second_format in formats:
        A = scipy.sparse.eye(nA, nB, format=first_format)
        B = scipy.sparse.eye(nB, nB, format=second_format)
        sparse_dot_mkl.dot_product_mkl(A, B, dense=True)
        mem_start = memory_profiler.memory_usage()[0]
        start = time.time()
        for i in range(10000):
            sparse_dot_mkl.dot_product_mkl(A, B, dense=True)
        end = time.time()
        mem_end = memory_profiler.memory_usage()[0]
        print(f'{int(mem_end-mem_start)} MiB leaked in {first_format} x {second_format} multiplication (taking {int((end-start)*100)/100} seconds)')

print('\nversions')
import pkg_resources
import sys
print('sparse-dot-mkl', pkg_resources.get_distribution('sparse-dot-mkl').version)
print('scipy', pkg_resources.get_distribution('scipy').version)
print('numpy', pkg_resources.get_distribution('numpy').version)
print('mkl', sparse_dot_mkl.get_version_string())
print('python', sys.version)
```

prints

```0 MiB leaked in dense x dense multiplication (taking 2.99 seconds)
0 MiB leaked in dense x csr multiplication (taking 1.66 seconds)
575 MiB leaked in dense x csc multiplication (taking 2.35 seconds)
0 MiB leaked in csr x dense multiplication (taking 1.49 seconds)
0 MiB leaked in csr x csr multiplication (taking 1.91 seconds)
0 MiB leaked in csr x csc multiplication (taking 2.07 seconds)
0 MiB leaked in csc x dense multiplication (taking 3.17 seconds)
0 MiB leaked in csc x csr multiplication (taking 2.3 seconds)
0 MiB leaked in csc x csc multiplication (taking 2.17 seconds)

versions
sparse-dot-mkl 0.7.2
scipy 1.6.3
numpy 1.19.4
mkl Intel(R) oneAPI Math Kernel Library Version 2021.2-Product Build 20210312 for Intel(R) 64 architecture applications
python 3.8.8 (default, Apr 13 2021, 19:58:26) 
[GCC 7.3.0]
```

A quick workaround would be to call `.tocsr` on `csc` matrix inputs in this combination [seems to be a little faster, too]. But of course it would be nice, if the leak did not occur in the first place.

**Comment by asistradition** [comment]
Good catch - MKL doesn't support CSC 100% and the memory leak was in a case where it has to make an internal conversion to CSR (it's faster to do that in MKL than in scipy). I cleaned it up in 839d4a960163dc6876fd133729efe03d87896216 and I'll package it and release it today.

**Comment by asistradition** [comment]
Packaged and released as v0.7.3 - let me know if there are any additional problems.

Thanks

**Comment by simonwm** [comment]
Really amazing: both the package and the speed in which you resolve issues! Thanks a lot!
